### PR TITLE
refactor(cardinal): upstream config, refactor NewWorld, and replace WorldID with namespace

### PR DESCRIPTION
**Pull request by smsunarto**
Closes: WORLD-525

## What is the purpose of the change

Cleaning up the DevEx for creating a new world by upstreaming all the env var loading to upstream Cardinal SDK. Additionally, this PR also reduces the surface area of program entrypoint to make debugging/testing much more predictable.

## Brief Changelog

- Load cardinal related env var directly from `NewWorld`, upstreaming config from starter-game-template as a result
- Move `cardinal.NewTestWorld` to `testutils.NewTestWorld`
- Delete `ecs.NewTestWorld` and `ecs.NewMockWorld` to consolidate program entrypoint
- Replace `WorldID` with namespace that can be set through env var
- Remove options `WithPorts` and `WithNameserver` because they are supposed to be set through env var now
- Delete cardinal/LICENSE
- Delete cardinal/.gitignore

## Testing and Verifying

- Tests have been updated to accommodate the new flow.

### CODE REVIEW

**Comment by smsunarto** [comment]
Current dependencies on/for this PR:
* main
  * **PR #412** <a href="https://app.graphite.dev/github/pr/Argus-Labs/world-engine/412" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/Argus-Labs/world-engine/412?utm_source=stack-comment).

File: cardinal/config.go [hunk 1]
```diff
@@ -0,0 +1,28 @@
+package cardinal
+
+import "os"
+
+type WorldConfig struct {
+	RedisAddress       string
+	RedisPassword      string
+	CardinalWorldId    string
```
Comment (technicallyty) [inline cardinal/config.go:8]: Should this just be `CardinalNamespace` and have the optional namespace be removed?

File: cardinal/config.go [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+	CardinalPort       string
+	CardinalDeployMode string
```
Comment (technicallyty) [inline cardinal/config.go:10]: DeployMode feels like the wrong word here, as this doesn't control any sort of "deployment", just the type of world you get. Maybe just `Mode`? not really sure here.

either way ~
the DeployMode type should be more opinionated. string is too vague.

````go
type DeployMode string

const (
    DeployModeProd =  "produdction"
    DeployModeDev = "development"
   ....
)
```

File: cardinal/config.go [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+}
+
+func GetWorldConfig() WorldConfig {
+	return WorldConfig{
+		RedisAddress:       getEnv("REDIS_ADDRESS", "localhost:6379"),
+		RedisPassword:      getEnv("REDIS_PASSWORD", ""),
+		CardinalWorldId:    getEnv("CARDINAL_WORLD_ID", "world"),
+		CardinalPort:       getEnv("CARDINAL_PORT", "3333"),
```
Comment (technicallyty) [inline cardinal/config.go:18]: the default cardinal port is 4040. are we doing 3333 or 4040? if we want to go with 3333, we should update the docs and comments in this PR. additionally, in cardinal/server/server.go in the Initialize method, we load this env variable again. that should be removed if we're doing this here.

File: cardinal/world.go [hunk 4]
```diff
@@ -54,22 +54,27 @@ type (
...
+		log.Logger.Info().Msg("Starting a new Cardinal world in production mode")
+		if cfg.RedisPassword == "" {
+			return nil, errors.New("redis password is required in production")
+		}
+	} else {
+		if cfg.CardinalDeployMode != "development" {
+			log.Logger.Warn().Msg("CARDINAL_DEPLOY_MODE is unrecognized. Defaulting to development mode")
+		}
+		log.Logger.Info().Msg("Starting a new Cardinal world in development mode")
+	}
```
Comment (technicallyty) [inline cardinal/world.go:72]: not really sure what this control flow is supposed to control. Is "development" supposed to start world with miniredis?

File: cardinal/world.go [hunk 5]
```diff
@@ -93,22 +98,12 @@ func NewWorld(addr, password string, opts ...WorldOption) (*World, error) {
 	return world, nil
 }
 
-// NewMockWorld creates a World that uses an in-memory redis DB as the storage layer.
-// This is only suitable for local development.
+// NewMockWorld creates a World object that uses miniredis as the storage layer suitable for local development.
+// If you are creating an ecs.World for unit tests, use NewTestWorld.
```
Comment (technicallyty) [inline cardinal/world.go:102]:
```suggestion
// If you are creating a World for unit tests, use NewTestWorld.
```

File: cardinal/config.go (hunk 1 above)
Comment (smsunarto) [reply cardinal/config.go:8]: I agree. I can do that

File: cardinal/config.go (hunk 2 above)
Comment (smsunarto) [reply cardinal/config.go:10]: I'm down to just call it CardinalMode and using enum

File: cardinal/config.go (hunk 3 above)
Comment (smsunarto) [reply cardinal/config.go:18]: Didn't realize we decided on 4040. Will update this and World CLI accordingly

File: cardinal/world.go (hunk 4 above)
Comment (smsunarto) [reply cardinal/world.go:72]: Dev mode just doesn't enforce redis password and I wanted to set the logger to pretty print as well but I forgot

File: cardinal/world.go [hunk 6]
```diff
@@ -93,22 +98,12 @@ func NewWorld(addr, password string, opts ...WorldOption) (*World, error) {
...
-	world := &World{
-		implWorld:     implWorld,
-		serverOptions: serverOptions,
-		cleanup:       mockWorldCleanup,
-		endStartGame:  make(chan bool),
-	}
-	world.isGameRunning.Store(false)
-	for _, opt := range cardinalOptions {
-		opt(world)
+	world, err := NewWorld(append(opts, withMockRedis())...)
```
Comment (technicallyty) [inline cardinal/world.go:104]:
```suggestion
	world, err := NewWorld(withMockRedis(), opts...)
```

Comment (smsunarto) [reply cardinal/world.go]: this doesn't work btw

File: cardinal/world.go [hunk 7]
```diff
@@ -263,32 +276,36 @@ func RegisterSystems(w *World, systems ...System) {
...
 }
 
 // RegisterQueries adds the given query capabilities to the game world. HTTP endpoints to use these queries
 // will automatically be created when StartGame is called. This Register method must only be called once.
 func RegisterQueries(w *World, queries ...AnyQueryType) error {
-	return w.implWorld.RegisterQueries(toIQueryType(queries)...)
+	return w.instance.RegisterQueries(toIQueryType(queries)...)
+}
+
+func (w *World) Instance() *ecs.World {
```
Comment (jerargus) [inline cardinal/world.go:294]: This is meant to only be used in tests, correct? I'd suggest renaming this to TestInstance or TestToECSWorld, and adding an appropriate comment.

File: cardinal/config.go [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+import (
+	"os"
+
+	"github.com/rs/zerolog/log"
+)
+
+const (
+	CardinalModeProd         = "production"
+	CardinalModeDev          = "development"
+	DefaultCardinalNamespace = "world-1"
```
[the remark this answers is not in the document]
Comment (jerargus) [reply cardinal/config.go:12]: +1

**Review by jerargus — APPROVED** [review]

File: cardinal/world.go (hunk 7 above)
Comment (smsunarto) [reply cardinal/world.go:294]: Not necessarily only for test, but I'll just call it InternalInstance to make it clear that people should typically not use this

File: cardinal/config.go [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+	DefaultRedisPassword     = ""
+)
+
+type WorldConfig struct {
+	RedisAddress      string
+	RedisPassword     string
+	CardinalNamespace string
+	CardinalPort      string
+	CardinalMode      string
```
[the remark this answers is not in the document]
Comment (smsunarto) [reply cardinal/config.go:21]: would rather not, these are env var and it's a chore to keep typecasting it

File: cardinal/world.go (hunk 7 above)
Comment (smsunarto) [reply cardinal/world.go:295]: you know what, let's make a separate ticket for this. the cascading rebase i have to do for this is not very pleasant.

**Comment by smsunarto** [comment]
### Merge activity

* **Nov 13, 3:25 PM**: [Graphite](https://app.graphite.dev/github/pr/Argus-Labs/world-engine/412) rebased this pull request as part of a merge.

**Pull request merged**